src/Adapter/ContextStateManager.php

```diff
@@ -151,11 +159,37 @@ public function setCustomer(?Customer $customer): self
      *
      * @return self
      */
-    public function restoreContext(): self
+    public function restorePreviousContext(): self
```

Comment:
How do you know what state are you restoring to?

```php
$this->contextMgr->saveCurrentContext();
$this->contextMgr->setCustomer($newCustomer);
doSomething(); // changes context but does not restore it after for some reason
$this->contextMgr->restorePreviousContext(); // state is not restored as expected
```

You need to keep track of snapshots instead:
```php
$snapshotId = $this->contextMgr->saveSnapshot(); // returns snapshot id that never changes
doSomething(); // stuff happens
$this->contextMgr->restore($snapshotId); // you know exactly what you're coming back to
```

Reply:
We are aware there are edge cases where the stash / pop strategy can go wrong if used by users unaware of the inner mechanics behind this service.

We have listed some of them here https://github.com/PrestaShop/PrestaShop/issues/21579 but we feel like it could be done for 1.7.8

Reply:
If the context is modified directly not using the ContextStateManager we can't know them and restore them, so we can do nothing for this one..

If it is modified via the ContextStateManager we only save the first value that has been overridden, following calls to `setCurrency` (for example) will be applied but not saved So when we restore the context it is the original value (or previous) value that is restored

I just added a unit test to be sure of this behaviour and to convince you better ^^ https://github.com/PrestaShop/PrestaShop/blob/43da1d65830e2fee71a12b533228f3e958922917/tests/Unit/Adapter/ContextStateManagerTest.php#L213

Regarding the `$snapshotId` I lie the idea but I think it could be added in another version, for now we wanted a simple API push/pop contexts in a stack without having to store an ID This makes the use easier I think

Reply:
I think indeed the idea of snapshots is different from what we are doing here.

There 2 needs, and 2 usecases that sometimes overlap:

**Usecase 1**

I know I am going to override `$context->currency` because _I need to do it_. Maybe I am modifying some cart property and I know it relies on context currency. So I use ContextStateManager to restore this property after I am done. It's like "hold my beer, I'm fixing this stuff then you give it back to me".

So I know what property of the `$context` I am messing with, and I do it voluntarily.

**Usecase 2**

I know I am going to perform some operations that I expect to modify `$context` properties but I dont know which ones exactly. I want to make sure that, once these operations are done, I restore the context to its previous state, whatever it was.

So I dont know what property of the `$context` I am going to mess with, and it happens against my will, because the code I will execute is doing that.